# Lista 8 (4pkt.+5pkt.) Termin: 14.01.2020r.

In [1]:
import numpy as np
import random


In [2]:
def pow_mod(x, n, m):
    bits = format(int(n), '0b')
    p = 1
    for i in range(len(bits)):
        p = p * p % m
        if bits[i] == "1":
            p = p * x % m
    return p

# Zadanie 1 (2 pkt.)

Alicja i Bob ustalają publicznie grupę $Z_p$ w której będą działać oraz element z tej grupy $g$.

In [3]:
g=5
p=23

Alicja wybiera losowy element $a$ grupy $Z_p$, taki, że $1<a<p$:

In [4]:
a = random.randint(1, p)
print(a)

19


Następnie oblicza $A=g^a$ w grupie $Z_p$, czyli $A=g^a\mod\ p$:

In [5]:
A=pow_mod(g,a,p)
print(A)

7


To samo robi Bob, wybiera losowy elemnt grupy $b$ i oblicza $B=g^b$:

In [6]:
b= random.randint(1, p)
B= pow_mod(g,b,p)
print(b)
print(B)

16
3


Elementy $a$ i $b$ są znane tylko, odpowiednio, Alicji i Bobowi, natomiast $A$ oraz $B$ są wymieniane publicznym kanałem. Bob otrzymuje $A$ i wykonuje operację $k_B=A^b$ (czyli $A^b\mod\ p$), z kolei Alicja otrzymuje $B$ i wykonuje operację $k_A=B^a$ (czyli $A^b\mod\ p$):

In [7]:
k_B= pow_mod(A, b, p)
k_A= pow_mod(B, a, p)
print(k_B)
print(k_A)

6
6


Zawuażmy, że $k_A=k_B$, wynika to z tego, że $k_A=B^a=(g^b)^a=g^{ab}$ oraz $k_B=A^b=(g^a)^b=g^{ba}$ a oczywiście $ab=ba$. Zatem Alicja i Bob dysponują tą samą liczbą, którą mogą użyć jako klucz do szyfrowania. Publicznie znane są wartości $g$, $p$, $g^a$ oraz $g^b$, jednak odzyskanie z nich $a$ i $b$ jest bardzo trudne przy odpwiednim doborze grupy $Z_p$.

## Zadanie 2 (2 pkt.)

Wczytaj dwa obrazki: 'Campin_Boze_Narodzenie1.png' oraz 'Campin_Boze_Narodzenie2.png' jeden z nich jest zgodny z oryginałem obrazu Roberta Campin a drugi jest lekko zmodyfikowany, rozponaj prawdziwy obraz obliczając jego funkcję skrótu SHA-256 i porównując ją do skrótu podanego poniżej (w postaci szesnastkowej), który odpowiada oryginałowi:

In [8]:
hash_original='2d08246bbb3a8f6b0c16b8d8effd2c783233a6a174496731dfff887bb41bc7e7'

Obrazek wczytaj za pomocą:

In [9]:
file = open('Campin_Boze_Narodzenie1.png','rb')
obraz1=file.read()
file.close()
file = open('Campin_Boze_Narodzenie2.png','rb')
obraz2=file.read()
file.close()

In [12]:
h = hashlib.sha256()
h.update(obraz1)
hash_obraz1 = h.hexdigest()
print(hash_obraz1 == hash_original)

True


In [ ]:
Obraz 1 jest prawdzwy

In [13]:
k = hashlib.sha256()
k.update(obraz2)
hash_obraz2 = k.hexdigest()
print(hash_obraz2 == hash_original)

False


'rb' w powyższej funkcji oznacza, że odczytujemy binarną postać obrazu. Podobnie wczytaj drugi obraz. Do policzenia funkcji skrótu użyj biblioteki hashlib: https://docs.python.org/3/library/hashlib.html

In [11]:
import hashlib

## Zadanie 3 * (3 pkt.)

Na poprzednich listach pojawiała się funckja przesuwająca zadany ciag bitów o daną liczbę miejsc w lewo, np.:

In [10]:
def shift_left(tab,n):
    return tab[n:]+tab[:n]

In [11]:
shift_left('11101',2)

'10111'

Funkcja ta jednak działa na ciągu znaków tekstowych a nie bitach, napisz tą funkcję jeszcze raz tak aby działała na int-ach reprezentujących ciagi bitów:

In [14]:
def Shift_left(k,n,n_bits):
    return ((k<<n) & ((1<<n_bits) -1)) | k >> (n_bits - n)


tzn. powyższy przykład wyglądałby następująco:

In [15]:
print(Shift_left(29,2,5)==23)

True


bo '11101'=29 oraz '10111'=23, natomiast n_bits=5, oznacza, że ciąg bitów ma dlugość 5. W implementacji użyj operatorów takich jak: <<, |, >>, & (https://www.geeksforgeeks.org/python-bitwise-operators/)

## Zadanie 4 * (2 pkt.)

Na wykładzie byla pokazana metoda szybszej deszfyracji w RSA (zmiejszenia wykładnika w potęgowaniu):
$$M=(C^{d_q}+hq)\mod n$$
na wykładzie był podany wzór bez $\mod n$, jednak jest ono potrzebne, n=pq, pozostale oznaczenia jak na wykładzie (notatki na Teamsach), zaimplementuj powyższy wzór.